pkg/auth/auth.go

```diff
@@ -1,66 +1,137 @@
 package auth
 
 import (
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
 	"fmt"
+	"net/url"
 	"regexp"
-	"strings"
 
 	"github.com/xenitab/azdo-proxy/pkg/config"
 )
 
+const tokenLenght = 64
+
 type Authorization struct {
-	Endpoints []Endpoint
+	endpoints map[string]Endpoint
 }
 
 type Endpoint struct {
-	Token   string
-	Regexes []*regexp.Regexp
-}
+	Pat        string
+	Domain     string
+	Scheme     string
+	Token      string
+	Namespaces []string
+	SecretName string
 
-// GenerateAuthorization creates regex resources to validate tokens and endpoints
-func GenerateAuthorization(c config.Configuration) (*Authorization, error) {
-	baseApi, err := regexp.Compile(`/` + c.Organization + `/_apis\b`)
-	if err != nil {
-		return nil, err
-	}
+	// metadata used for reverse lookup
+	Organization string
+	Project      string
+	Repository   string
 
-	auth := &Authorization{Endpoints: []Endpoint{}}
-	for _, r := range c.Repositories {
-		git, err := regexp.Compile(`/` + c.Organization + `/` + r.Project + `/_git/` + r.Name + `(/.*)?\b`)
-		if err != nil {
-			return nil, err
-		}
+	regexes []*regexp.Regexp
+}
 
-		api, err := regexp.Compile(`/` + c.Organization + `/` + r.Project + `/_apis/git/repositories/` + r.Name + `(/.*)?\b`)
+// NewAuthorization cretes a new authorization from a give configuration.
+func NewAuthorization(cfg config.Configuration) (Authorization, error) {
+	authz := Authorization{endpoints: map[string]Endpoint{}}
+	for _, o := range cfg.Organizations {
+		baseApi, err := regexp.Compile(fmt.Sprintf(`/%s/_apis\b`, o.Name))
 		if err != nil {
-			return nil, err
+			return Authorization{}, fmt.Errorf("invalid base api regex: %v", err)
 		}
 
-		endpoint := Endpoint{
-			Token:   r.Token,
-			Regexes: []*regexp.Regexp{baseApi, git, api},
+		for _, r := range o.Repositories {
+			git, err := regexp.Compile(fmt.Sprintf(`/%s/%s/_git/%s(/.*)?\b`, o.Name, r.Project, r.Name))
+			if err != nil {
+				return Authorization{}, err
+			}
+			api, err := regexp.Compile(fmt.Sprintf(`/%s/%s/_apis/git/repositories/%s(/.*)?\b`, o.Name, r.Project, r.Name))
+			if err != nil {
+				return Authorization{}, err
+			}
+			token, err := randomSecureToken()
+			if err != nil {
+				return Authorization{}, fmt.Errorf("could not generate random token: %v", err)
+			}
+
+			e := Endpoint{
+				Pat:          o.Pat,
+				Domain:       o.Domain,
+				Scheme:       o.Scheme,
+				Token:        token,
+				Namespaces:   r.Namespaces,
+				SecretName:   o.GetSecretName(r),
+				Organization: o.Name,
+				Project:      r.Project,
+				Repository:   r.Name,
+				regexes:      []*regexp.Regexp{baseApi, git, api},
+			}
+			authz.endpoints[token] = e
 		}
-		auth.Endpoints = append(auth.Endpoints, endpoint)
 	}
 
-	return auth, nil
+	return authz, nil
 }
 
-// IsPermitted checks if a specific user is permitted to access a path
-func IsPermitted(a *Authorization, path string, token string) error {
```

Comment:
This may be more of a philosophical question, but in regards to the `ok` vs. `found` comment using the Go way - shouldn't we have a Token struct and a `Valid()` method on it like the oauth lib?

```Go
func (t *Token) Valid() bool
```